src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp

```diff
@@ -1130,63 +824,72 @@ namespace
             const size_t SampleCount = 512;
 
             for (size_t i = 0; i < SampleCount; ++i)
             {
-                // Generate a uniform sample in [0,1)^3.
-                const size_t Bases[] = { 2, 3 };
-                const Vector3f s = hammersley_sequence<float, 3>(Bases, SampleCount, i);
-
-                Vector3f wi;
-                float value = 0.0f;
-                float probability = 0.0f;
-
-                GlassBSDFImpl::do_sample(
-                    mdf,
-                    s,
-                    false,
-                    shading_basis,
-                    alpha,
-                    alpha,
-                    1.0f, // reflection_color
-                    1.0f, // reflection_weight
-                    1.0f, // refraction_color
-                    1.0f, // refraction_weight
-                    eta,
-                    wo,
-                    wi,
-                    value,
-                    probability);
-                assert(probability >= 0.0f);
+                // Generate a uniform sample in [0,1)^2.
+                const size_t Bases[] = { 2 };
+                const Vector2f s = hammersley_sequence<float, 2>(Bases, SampleCount, i);
+
+                // Compute the microfacet normal by sampling the MDF.
+                const Vector3f m = GGXMDF::sample(wo, s, alpha, alpha, 0.0f);
+                assert(m.y > 0.0f);
+
+                // Compute the Fresnel term.
+                const float rcp_eta = 1.0f / eta;
+                const float cos_wom = clamp(dot(wo, m), -1.0f, 1.0f);
+
+                float cos_theta_t;
+                const float F = fresnel_reflectance(cos_wom, rcp_eta, cos_theta_t);
 
-                if (probability < 1.0e-6f)
-                    continue;
+                const float rcp_G1 = safe_rcp(GGXMDF::G1(wo, m, alpha, alpha, 0.0f));
+
+                // Evaluate the reflection lobe.
+                {
+                    const Vector3f wi = improve_normalization(reflect(wo, m));
+
+                    if (wi.y * wo.y > 0.0f)
+                    {
+                        const float G = GGXMDF::G(wi, wo, m, alpha, alpha, 0.0f);
+                        R += F * G * rcp_G1;
+                    }
+                }
 
-                R += value * abs(wi.y) / probability;
+                // Evaluate the transmission lobe.
+                {
+                    const Vector3f wi = refracted_direction(
+                        wo,
+                        m,
+                        cos_wom,
+                        cos_theta_t,
+                        rcp_eta);
+
+                    if (wi.y * wo.y <= 0.0f)
+                    {
+                        const float G = GGXMDF::G(wi, wo, m, alpha, alpha, 0.0f);
+                        R += (1.0f - F) * G * rcp_G1;
+                    }
+                }
             }
 
             return min(R / static_cast<float>(SampleCount), 1.0f);
         }
+
+        static float safe_rcp(const float x)
```

Comment:
This could be moved to `foundation/math/scalar.h` together with the other `safe_*()` methods. The epsilon value should be passed in argument and it should have `default_eps<T>()` as a default value.

Reply:
I'll include it in the next PR